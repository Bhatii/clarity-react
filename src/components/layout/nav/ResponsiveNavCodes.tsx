export class ResponsiveNavCodes {
    public static NAV_LEVEL_1: number = 1;
    public static NAV_LEVEL_2: number = 2;
    public static NAV_TYPE_HEADER: string = "NAV_TYPE_HEADER";
    public static NAV_TYPE_SIDE: string = "NAV_TYPE_SIDE";
    public static NAV_TYPE_SUB: string = "NAV_TYPE_SUB";
    public static NAV_OPEN: string = "NAV_OPEN";
    public static NAV_CLOSE: string = "NAV_CLOSE";
    public static NAV_TOGGLE: string = "NAV_TOGGLE";
    public static NAV_CLOSE_ALL: string = "NAV_CLOSE_ALL";
    public static NAV_CLASS_HAMBURGER_MENU: string = "open-hamburger-menu";
    public static NAV_CLASS_OVERFLOW_MENU: string = "open-overflow-menu";
}
