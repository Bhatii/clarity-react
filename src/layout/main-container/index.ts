export * from "./MainContainer";
export * from "./ClassNames";
