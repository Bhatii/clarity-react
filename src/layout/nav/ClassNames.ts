export class ClassNames {
    public static Actions: string = "header-actions";
    public static HamburgerTrigger: string = "header-hamburger-trigger";
    public static HamburgerMenu: string = "open-hamburger-menu";
    public static Header1: string = "header-1";
    public static Header2: string = "header-2";
    public static Header3: string = "header-3";
    public static Header4: string = "header-4";
    public static Header5: string = "header-5";
    public static Header6: string = "header-6";
    public static Header7: string = "header-7";
    public static OverflowTrigger: string = "header-overflow-trigger";
    public static OverflowMenu: string = "open-overflow-menu";
    public static Search: string = "search";
}
