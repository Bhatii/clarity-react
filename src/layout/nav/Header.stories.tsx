import * as React from 'react';
import {storiesOf} from '@storybook/react';
import {Header} from ".";

storiesOf('Header', module)
    .add('Header', () =>
        <Header isNavLevel1OnPage={true} isNavLevel2OnPage={true}/>
    )
;
