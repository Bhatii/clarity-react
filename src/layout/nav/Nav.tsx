import * as React from "react";
import * as utils from '../../utils';

export type NavProps = {
    // The level of navigation invoked.
    navLevel: NavLevel;

    // The navigation type (header, sub side).
    navType: NavType;
};

export enum NavLevel {
    PRIMARY = "1",
    SECONDARY = "2"
}

export enum NavType {
    HEADER = "NAV_TYPE_HEADER",
    SIDE = "NAV_TYPE_SIDE",
    SUB = "NAV_TYPE_SUB"
}

export class Nav extends React.PureComponent<NavProps> {
    private static classForNavType(navType: NavType): string {
        switch (navType) {
            case NavType.HEADER:
                return "header-nav";
            case NavType.SIDE:
                return "sidenav";
            case NavType.SUB:
                return "subnav";
            default:
                return "header-nav";
        }
    }

    private static classForNavLevel(navLevel: NavLevel): string {
        return "clr-nav-level=" + navLevel;
    }

    render() {
        const {children, navLevel, navType} = this.props;
        const classList: string[] = [
            Nav.classForNavType(navType),
            Nav.classForNavLevel(navLevel)
        ];
        return (
            <nav className={utils.classNames(classList)}>
                {children}
            </nav>
        );
    }
}
