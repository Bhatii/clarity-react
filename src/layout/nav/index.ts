export * from "./Header";
export * from "./Nav";
export * from "./NavLink";
export * from "./ClassNames";
