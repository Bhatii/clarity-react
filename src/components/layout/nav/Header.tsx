import * as React from "react";
import {ResponsiveNavCodes} from "./ResponsiveNavCodes";

export type HeaderProps = {
    // onToggle handles any calls made in this component that will trigger
    // navigation toggling. The logic for handling the toggle is within the
    // parent object, and will have an effect on the navList property.
    onToggle: (controlCode: string, navLevel: number) => void;

    // onCloseAll handles a call made to close all in this component. The logic
    // for handling the closeAll action is within the parent object, and will
    // have an effect on navList property.
    onCloseAll: () => void;

    // navList contains a list of numbers representing the active navigation
    // levels.
    navList: number[];
};

type navStates = {
    isNavLevel1OnPage: boolean;
    isNavLevel2OnPage: boolean;
};

export default class Header extends React.PureComponent<HeaderProps> {
    state: navStates = {
        isNavLevel1OnPage: false,
        isNavLevel2OnPage: false,
    };

    constructor(props: HeaderProps) {
        super(props);
        this.state = this.initializeNavTriggers(props);
    }

    // initializeNavTriggers sets the navigation state for this component based
    // on whether each navLevel is contained within the navList
    private initializeNavTriggers(props: HeaderProps): navStates {
        const {NAV_LEVEL_1, NAV_LEVEL_2} = ResponsiveNavCodes;
        let isNavLevel1OnPage = false,
            isNavLevel2OnPage = false;
        if (props.navList.length > 2) {
            console.error("More than 2 Nav Levels detected.");
            return {isNavLevel1OnPage, isNavLevel2OnPage};
        }
        props.navList.forEach(navLevel => {
            if (navLevel === NAV_LEVEL_1) {
                isNavLevel1OnPage = true;
            } else if (navLevel === NAV_LEVEL_2) {
                isNavLevel2OnPage = true;
            }
        });
        return {isNavLevel1OnPage, isNavLevel2OnPage};
    }

    handleToggle(navLevel: number) {
        const {NAV_TOGGLE} = ResponsiveNavCodes;
        this.props.onToggle(NAV_TOGGLE, navLevel);
    }

    render() {
        const {onCloseAll, onToggle} = this.props;
        const {NAV_LEVEL_1, NAV_LEVEL_2} = ResponsiveNavCodes;
        const {isNavLevel1OnPage, isNavLevel2OnPage} = this.state;

        return (
            <header className="header">
                {isNavLevel1OnPage && (
                    <button className="header-hamburger-trigger" type="button" onClick={this.handleToggle.bind(this, NAV_LEVEL_1)}>
                        <span />
                    </button>
                )}
                {this.props.children}
                {isNavLevel2OnPage && (
                    <button className="header-overflow-trigger" type="button" onClick={this.handleToggle.bind(this, NAV_LEVEL_2)}>
                        <span />
                    </button>
                )}
                <div className="header-backdrop" onClick={onCloseAll} />
            </header>
        );
    }
}
