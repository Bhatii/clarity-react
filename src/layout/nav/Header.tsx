import * as React from "react";
import {ResponsiveNavCodes} from "./ResponsiveNavCodes";

export type HeaderProps = {
    // navList is a list of numbers representing the navigation components enabled
    // in the pane
    navList?: number[];

    // onHamburgerToggle handles toggle actions for the left side navigation
    // "hamburger" icon"
    onHamburgerToggle?: () => void;

    // onRightSideToggle handles toggle actions for the left side navigation
    // vertical ellipsis icon
    onRightSideToggle?: () => void;

    // onCloseAll handles a call made to close all in this component. The logic
    // for handling the closeAll action is within the parent object, and will
    // have an effect on navList property.
    onCloseAll?: () => void;
};

type navStates = {
    isNavLevel1OnPage: boolean;
    isNavLevel2OnPage: boolean;
};

export class Header extends React.PureComponent<HeaderProps> {
    state: navStates = {
        isNavLevel1OnPage: false,
        isNavLevel2OnPage: false,
    };

    constructor(props: HeaderProps) {
        super(props);
        this.state = this.initializeNavTriggers(props);
    }

    // initializeNavTriggers sets the navigation state for this component based
    // on whether each navLevel is contained within the navList
    private initializeNavTriggers(props: HeaderProps): navStates {
        if (!props.navList)
            return {isNavLevel1OnPage: false, isNavLevel2OnPage: false};
        const {NAV_LEVEL_1, NAV_LEVEL_2} = ResponsiveNavCodes;
        let isNavLevel1OnPage = false,
            isNavLevel2OnPage = false;
        if (props.navList && props.navList.length > 2) {
            console.error("More than 2 Nav Levels detected.");
            return {isNavLevel1OnPage, isNavLevel2OnPage};
        }
        props.navList.forEach(navLevel => {
            if (navLevel === NAV_LEVEL_1) {
                isNavLevel1OnPage = true;
            } else if (navLevel === NAV_LEVEL_2) {
                isNavLevel2OnPage = true;
            }
        });
        return {isNavLevel1OnPage, isNavLevel2OnPage};
    }

    render() {
        const {onCloseAll, onHamburgerToggle, onRightSideToggle} = this.props;
        const {isNavLevel1OnPage, isNavLevel2OnPage} = this.state;

        return (
            <header className="header">
                {isNavLevel1OnPage && (
                    <button
                        className="header-hamburger-trigger"
                        type="button"
                        onClick={onHamburgerToggle}
                    >
                        <span />
                    </button>
                )}
                {this.props.children}
                {isNavLevel2OnPage && (
                    <button
                        className="header-overflow-trigger"
                        type="button"
                        onClick={onRightSideToggle}
                    >
                        <span />
                    </button>
                )}
                <div className="header-backdrop" onClick={onCloseAll} />
            </header>
        );
    }
}
