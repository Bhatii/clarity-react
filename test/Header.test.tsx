import {Header} from "../src/layout/nav";
import {shallow} from "enzyme";
import * as React from "react";

let wrapper: any;

beforeEach(() => {
    wrapper = shallow(
        <Header
            navList={[]}
            onCloseAll={() => {}}
            onHamburgerToggle={() => {}}
            onRightSideToggle={() => {}}
        />,
    );
});

describe("<MainContainer /> rendering", () => {
    it("renders correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });
    it("should not have any buttons", () => {
        expect(wrapper.find("button")).toHaveLength(0);
    });
    // it("should have one button for hamburger", () => {
    //     wrapper.setProps({navList: [1, 0]});
    //     expect(wrapper.find("button")).toHaveLength(1);
    // });
    // it("should have one button for overflow", () => {
    //     wrapper.setProps({navList: [0, 2]});
    //     expect(wrapper.find("button")).toHaveLength(1);
    // });
    it("should have two buttons for hamburger and overflow", () => {
        const test = shallow(
            <Header
                navList={[1, 2]}
                onCloseAll={() => {}}
                onHamburgerToggle={() => {}}
                onRightSideToggle={() => {}}
            />,
        );
        //test.setProps({navList: [1, 2]});
        expect(test.find("button")).toHaveLength(2);
    });
});
