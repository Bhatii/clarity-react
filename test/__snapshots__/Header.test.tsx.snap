// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainContainer /> rendering renders correctly 1`] = `
<header
  className="header"
>
  <div
    className="header-backdrop"
    onClick={[Function]}
  />
</header>
`;
